sympy/solvers/solveset.py

```diff
@@ -715,117 +525,120 @@ def _solve_abs(f, symbol):
                                            symbol).intersect(q_neg_cond)
         return Union(sols_q_pos, sols_q_neg)
     else:
-        return ConditionSet(symbol, Eq(f, 0), S.Complexes)
-
-
-def solveset_complex(f, symbol):
-    """ Solve a complex valued equation.
-
-    Parameters
-    ==========
-
-    f : Expr
-        The target equation
-    symbol : Symbol
-        The variable for which the equation is solved
-
-    Returns
-    =======
+        return ConditionSet(symbol, Eq(f, 0), domain)
 
-    Set
-        A set of values for `symbol` for which `f` equal to
-        zero. An `EmptySet` is returned if no solution is found.
-        A `ConditionSet` is returned as an unsolved object if algorithms
-        to evaluate complete solutions are not yet implemented.
-
-    `solveset_complex` claims to be complete in the solution set that
-    it returns.
-
-    Raises
-    ======
-
-    NotImplementedError
-        The algorithms to solve inequalities in complex domain  are
-        not yet implemented.
-    ValueError
-        The input is not valid.
-    RuntimeError
-        It is a bug, please report to the github issue tracker.
 
-    See Also
-    ========
 
-    solveset_real: solver for real domain
+def _solveset(f, symbol, domain, _final=False, _check=False):
+    """Helper for solveset to return a result from an expression
+    that has already been sympify'ed and is known to contain the
+    given symbol."""
+    # _final controls whether the final result is being returned
+    # or not; if so, _check controls whether the answer is
+    # checked or not.
 
-    Examples
-    ========
-
-    >>> from sympy import Symbol, exp
-    >>> from sympy.solvers.solveset import solveset_complex
-    >>> from sympy.abc import x, a, b, c
-    >>> solveset_complex(a*x**2 + b*x +c, x)
-    {-b/(2*a) - sqrt(-4*a*c + b**2)/(2*a), -b/(2*a) + sqrt(-4*a*c + b**2)/(2*a)}
-
-    * Due to the fact that complex extension of my real valued functions are
-      multivariate even some simple equations can have infinitely many
-      solution.
-
-    >>> solveset_complex(exp(x) - 1, x)
-    ImageSet(Lambda(_n, 2*_n*I*pi), Integers())
-
-    """
-    if not getattr(symbol, 'is_Symbol', False):
-        raise ValueError('A Symbol must be given, not type %s: %s' %
-            (type(symbol), symbol))
+    orig_f = f
+    f = together(f)
+    if f.is_Mul:
+        _, f = f.as_independent(symbol, as_Add=False)
+    if f.is_Add:
+        a, h = f.as_independent(symbol)
+        m, h = h.as_independent(symbol, as_Add=False)
+        f = a/m + h  # XXX condition `m != 0` should be added to soln
+    f = piecewise_fold(f)
 
-    f = sympify(f)
-    original_eq = f
-    if not isinstance(f, (Expr, Number)):
-        raise ValueError(" %s is not a valid sympy expression" % (f))
+    # assign the solvers to use
+    solver = lambda f, x, domain=domain: _solveset(f, x, domain)
```

Comment:
This is same as `solver = _solveset` also since there is not need to specify the solver for `_solve_as_poly`, you might as well do away with this line.